free/src/main/scala/cats/free/Inject.scala

```diff
@@ -0,0 +1,49 @@
+package cats.free
+
+import cats.Functor
+import cats.data.Coproduct
+
+
+/**
+  * Inject type class as described in "Data types a la carte" (Swierstra 2008).
+  *
+  * @see [[http://www.staff.science.uu.nl/~swier004/Publications/DataTypesALaCarte.pdf]]
+  */
+sealed abstract class Inject[F[_], G[_]] {
+  def inj[A](fa: F[A]): G[A]
+
+  def prj[A](ga: G[A]): Option[F[A]]
+}
+
+sealed abstract class InjectInstances {
```

Comment:
Oops sorry I wasn't very clear. The return types you added are nice, but I was thinking of the `implicit def`s. I think generally it's a good practice to put an explicit return type on all `implicit`s since it's a bit easy to accidentally have an overly specific type inferred.